### PR TITLE
(0.90.12) Bugfix for `SpitExplicitFreeSurface` on `MultiRegionGrid`s and make `SpitExplicitFreeSurface` the default + don't allow more halos than interior points

### LINKED ISSUE #3513: Change the default free-surface solver for `MultiRegionGrid`s to `SplitExplicitFreeSurface`
Currently the default is the `ImplicitFreeSurface`:

https://github.com/CliMA/Oceananigans.jl/blob/9f3f8e6d1469e570b0778f328ab864e879785ffd/src/MultiRegion/multi_region_grid.jl#L52-L54

### PR DESCRIPTION
**Pull request by navidcy**
This continues a bit #3503. As summarized below there seems to be a bug with `SpitExplicitFreeSurface` on `MultiRegionGrid`. This PR fixes this and then promotes `SpitExplicitFreeSurface` as the default free surface solver for `MultiRegionGrid`s.

Bugfix: `set!(::HydrostaticFreeSurfaceModel)` now calls `initialize!(model)` to set the barotropic transport that corresponds to the `u`, `v` velocities.

Closes #3513.

### CODE REVIEW

**Comment by navidcy** [comment]
Here's how the `u` field from the code-snippet above looks like when evolved on a SingleRegion and MultiRegion grids:

u

**Comment by simone-silvestri** [comment]
Hmmm okok looks like there's a problem there. It seems to me that also the tracer does not work is it right? Does it work with the implicit solver?

**Comment by navidcy** [comment]
Yes, the tracer works.
And yes, this works with the implicit solver.

This is essentially one of the tests from the `test_multiregion_advection_diffusion.jl`... and they pass on `main` using the `UnifiedImplicitSolver` with `MultiRegionGrid`.

**Comment by navidcy** [comment]
Perhaps there is an `@apply_regionally` that is missing from somewhere?
The fact that tracers are OK but velocities are not hints where the problem might be coming from?

**Comment by navidcy** [comment]
If you want we can Zoom and try to sort it out. No rush -- whenever.

**Comment by navidcy** [comment]
So @simone-silvestri, the tests now pass but I think it's totally wrong. It just happens that whatever wrong thing happens happens both in serial and multi-region....

```Julia
using Oceananigans, Logging, Test
using Oceananigans.Models.HydrostaticFreeSurfaceModels: update_state!
using Oceananigans.MultiRegion
using Oceananigans.MultiRegion: reconstruct_global_field

using GLMakie

Logging.disable_logging(Logging.Warn)

function initialize_model(grid; P = XPartition, regions, closure, field_name)

    model = HydrostaticFreeSurfaceModel(; grid = MultiRegionGrid(grid, partition = P(regions)),
                                          free_surface = SplitExplicitFreeSurface(substeps=5),
                                          closure,
                                          tracers = :c,
                                          coriolis = nothing,
                                          buoyancy=nothing)

    initial_condition(x, y, z) = cos(2x)

    f = fields(model)[field_name]

    @apply_regionally set!(f, initial_condition)

    update_state!(model)

    return model
end

Nx = Ny = 32

grid = RectilinearGrid(size = (Nx, Ny, 1), halo = (3, 3, 3),
                       topology = (Bounded, Bounded, Bounded),
                       x = (0, 1), y = (0, 1), z = (0, 1))

closure = ScalarDiffusivity(ν = 1, κ = 1)

field_name = :u

model_serial   = initialize_model(grid; P = XPartition, regions = 1, field_name, closure)
model_parallel = initialize_model(grid; P = XPartition, regions = 2, field_name, closure)

function plot_comparison(models, field_name, title)

    model_serial, model_parallel = models
    f_serial   = fields(model_serial)[field_name]
    f_parallel = reconstruct_global_field(fields(model_parallel)[field_name])

    fig = Figure()
    ax1 = Axis(fig[1, 1])
    ax2 = Axis(fig[2, 1])

    x, y, z = nodes(f_serial)
    hm1 = heatmap!(ax1, x, y, interior(f_serial, :, :, 1))
    Colorbar(fig[1, 2], hm1)

    x, y, z = nodes(f_parallel)
    hm2 = heatmap!(ax2, x, y, interior(f_parallel, :, :, 1))
    Colorbar(fig[2, 2], hm2)

    fig[0, :] = Label(fig, title, fontsize=20, tellwidth=false)

    return fig
end

fig = plot_comparison((model_serial, model_parallel), field_name, "before timestep")

save("before_timestep.png", fig)
```
before_timestep


```julia
Δt = 1e-6 * grid.Lz^2 / closure.κ

for _ = 1:1
    time_step!(model_serial, Δt)
    time_step!(model_parallel, Δt)
end

fig = plot_comparison((model_serial, model_parallel), field_name, "after 1 timestep")

save("after_1_timestep.png", fig)
```

after_1_timestep

File: src/Grids/input_validation.jl [hunk 1]
```diff
@@ -49,21 +64,30 @@ function validate_size(TX, TY, TZ, sz)
...
+# Note that if provided with halo=nothing, the default halo size for coord i
+# is the min(default_halo_size, size[i]).
+# While this is easy to change, many of tests might fail so this situation needs to be
 # cleaned up.
-function validate_halo(TX, TY, TZ, ::Nothing)
-    halo = Tuple(3 for i = 1:topological_tuple_length(TX, TY, TZ))
-    return validate_halo(TX, TY, TZ, halo)
+function validate_halo(TX, TY, TZ, size, ::Nothing)
+    default_halo_size = 3
+    halo = min.(size, Tuple(default_halo_size for _ = 1:3))
```
Comment (glwagner) [inline src/Grids/input_validation.jl:73]:
```suggestion
    maximum_halo = size
    default_halo = Tuple(default_halo_size for _ = 1:3)
    halo = map(min, default_halo, maximum_halo)
```

Comment (glwagner) [inline src/Grids/input_validation.jl:73]:
```suggestion
    maximum_halo = size
    default_halo = (3, 3, 3)
    halo = map(min, default_halo, maximum_halo)
```

File: src/MultiRegion/multi_region_grid.jl [hunk 2]
```diff
@@ -51,7 +51,7 @@ minimum_spacing(dir, grid::MultiRegionGrid, ℓx, ℓy, ℓz) =
 
 # the default free surface solver; see Models.HydrostaticFreeSurfaceModels
 default_free_surface(grid::MultiRegionGrid; gravitational_acceleration=g_Earth) =
-    ImplicitFreeSurface(; gravitational_acceleration)
+    SplitExplicitFreeSurface(; substeps=200, gravitational_acceleration)
```
Comment (simone-silvestri) [inline src/MultiRegion/multi_region_grid.jl:54]:
```suggestion
    SplitExplicitFreeSurface(; substeps=50, gravitational_acceleration)
```

File: src/Models/HydrostaticFreeSurfaceModels/split_explicit_free_surface_kernels.jl [hunk 3]
```diff
@@ -347,7 +347,7 @@ const FTS = FixedTimeStepSize
 # we set a minimum number of substeps to execute to avoid numerical issues
 const MINIMUM_SUBSTEPS = 5
 
-@inline calculate_substeps(substepping::FNS, Δt) = length(substepping.averaging_weights)
+@inline calculate_substeps(substepping::FNS, Δt=nothing) = length(substepping.averaging_weights)
```
Comment (simone-silvestri) [inline src/Models/HydrostaticFreeSurfaceModels/split_explicit_free_surface_kernels.jl:350]:
```suggestion
@inline calculate_substeps(substepping::FNS, Δt) = length(substepping.averaging_weights)
```

**Review by simone-silvestri — APPROVED** [review]

**Comment by navidcy** [comment]
the problem ended up being that with SplitExplicitFreeSurface the while we were initializing u or v we were not initializing the barotropic transport...

the PR adds the `initialize!(model)` in `set!(model, ...)` that does precisely that, i.e., ensures that the barotropic transport is set correctly to correspond to  u and v.

https://github.com/CliMA/Oceananigans.jl/blob/5e8b2f16d09475c124a2bef042cd7e13732eb9da/src/Models/HydrostaticFreeSurfaceModels/set_hydrostatic_free_surface_model.jl#L63

**Pull request merged**